### ISSUE TITLE
Is it possible to mark some variables as "Parameter" and only return gradients for them?

### ISSUE THREAD

**Issue by SunDoge** [google/jax#1983]
For example, we have `tf.Variable`, `torch.nn.Parameter`, `mxnet.gluon.Parameter`, `Zygote.Params` to mark the variables that require grad. Is it possible to do so in `jax`? It will make building neural networks much easier.

**Comment by shoyer** [comment]
You can't do this directly in JAX: `jax.grad()` returns a gradient with respect to function parameters. I'm sure this is possible to do with another layer on top of JAX, though.

**Comment by mattjj** [comment]
Well, there are no "variables" in JAX, so there's no way to mark them. But there are several ways to differentiate a function with respect to just some parameters.

For example, you can use [the `argnums` argument of `grad`](https://jax.readthedocs.io/en/latest/jax.html#automatic-differentiation). That tracks mathematical notation (from e.g. Spivak's classic _Calculus on Manifolds_); see [the Autodiff Cookbook](https://jax.readthedocs.io/en/latest/notebooks/autodiff_cookbook.html) for details.

You can also just use lexical closure or `functools.partial`:

```python
def f(x, y):
  ...

x = ...

grad_wrt_y = grad(lambda y: f(x, y))
grad_wrt_y = grad(partial(f, x))
```

That also tracks mathematical notation, where you can say something like "let g(y) = f(x, y), and consider ∂g".

Finally, if you want to build Variable / Parameter style semantics on top of JAX, you can, though it likely won't be functionally pure and hence won't be as composable. In general, functional components are easy to compose; we keep JAX and all its APIs functionally pure so that it's easier to build things out of them, and because it's only safe to transform pure function, but you can use state if you want so long as you call JAX transformations on pure functions.

WDYT?

**Comment by mattjj** [comment]
If you have a specific thing you're trying to express, and it's not clear how to do it with JAX's functional API, maybe you can post it here and we can think on it together?

**Comment by mattjj** [comment]
We used to have a brief explanation of [`jax.experimental.stax`](https://github.com/google/jax/blob/master/jax/experimental/README.md#neural-net-building-with-stax) in our main README, but we deleted it when streamlining things last December. I just added it back to [a sub-README](https://github.com/google/jax/blob/master/jax/experimental/README.md#neural-net-building-with-stax). I thought you might be interested because it's a small example of how to write neural networks in a functional style.

**Comment by SunDoge** [comment]
Since jax is in functional style, things will become different from pytorch and tf2.0 🤔. Thanks for answering.
